[topic-usb]First step to rework USB core.

eobalski opened the pull request:
usb: Rework the USB core internals.

Rework the way how the interfaces are handled by the USB core stack.
Until now USB stack was aware of only the first interface associated
with each class. This led to a problem when not first interface was
addressed by the request.

This patch allow to filter out the request at core level.
**This is not done in this PR but inits foreground**

The usb_cfg_data struct is legacy of initial implementation
of USB device stack. Initially the USB device stack was
implemented to support single class devices with
no composite support. This structure was intended to be used to
configure the device in run time by calling usb_set_config().

Because of how things grow the structure was used to represent
the classes instead of devices.

This patch removes device configuration descriptor pointer.
Adds pointer to table of USB interface containers associated with
he function it represents. Each container consist of pointer to
interface descriptor, its alternate setting (if exists) and currently
selected alternate setting. This help USB core stack to properly address
SET/GET interface requests.

This patch renames some of the `usb_cfg_data` fields to better
represent its meaning. The `usb_cfg_data` struct is renamed itself
as this name is no longer valid (look up).

**This patch is first step to rework the USB core stack and is not meant
to be considered as final form of USB core.**

Signed-off-by: Emil Obalski <emil.obalski@nordicsemi.no>

Commit:
usb: Remove pointer to device descriptor from usb_cfg_data

usb_cfg_data data structure is handling usb funciton specifig data.
USB device descriptor is single instance located in special RAM section.

This deletion is a step forward to make core stack aware of classes
it handles.

Signed-off-by: Emil Obalski <emil.obalski@nordicsemi.no>

Commit:
usb: Rename usb_cfg_data structure fields

Rename USB usb_cfg_data structures to better reflect its meaning.

interface -> request_handlers
    Name interface was used here as a communication interface
    between USB core stack and Class implementations.
    This naming could be missleading as USB specification defines
    interface as a part of the USB device descriptor.

endpoint -> endpoints
    This change was made to reflect that there might be more than one
    endpoint.

Signed-off-by: Emil Obalski <emil.obalski@nordicsemi.no>

tbursztyka commented on include/usb/usb_device.h:
```diff
@@ -183,15 +183,15 @@ struct usb_cfg_data {
 			      const uint8_t *param);
 	/** USB interface (Class) handler and storage space */
-	struct usb_interface_cfg_data interface;
+	struct usb_interface_cfg_data request_handlers;
```
Actually, would it make sens to rename struct usb_interface_cfg_data to struct usb_interface_handlers or something more relevant that "cfg_data".

emob-nordic replied:
Yes, that's right. I forgot to include that in this patch. `usb_interface_handlers` does not convince me 100%. I would like to avoid interface at all here as this could be misleading taking into consideration USB interface meaning.

emob-nordic replied:
I don't have better idea than naming it `usb_request_handlers`.

tbursztyka replied:
it's a better name yes

Commit:
usb: webusb: Use API structure instead of custom one

This patch cleans webusb class. WebUSB class were using custom
request handler structure which is not necessary. This one
is removed and API struct is used instead.

Signed-off-by: Emil Obalski <emil.obalski@nordicsemi.no>

Commit:
usb: Rename usb_cfg_data to usb_class_data.

The name 'usb_cfg_data' is legacy of initial implementation
of the USB device stack. This structure was intended to be used to
configure the device in run time by calling usb_set_config().

Because of how things grown this was used to represent the data
of classes rather than the device as a whole. One device may consist of
many classes (composite device). When the stack was initially
developed this was not considered.

Alongside with renaming usb_cfg_data -> usb_class_data, all relevant
functions/macros/variables are updated accordingly.

Signed-off-by: Emil Obalski <emil.obalski@nordicsemi.no>

jfischer-phytec-iot commented on include/usb/usb_device.h:
```diff
@@ -143,6 +143,20 @@ struct usb_ep_cfg_data {
+ * only one alternate setting, this may change in the future.
+ */
+struct usb_if_container {
+	const struct usb_if_descriptor *const iface;
```
This well be reworked in subsequent patches as discussed, right? (just to clarify)

rarok59 replied:
Yes. I don't want to make everything in one PR, as it will grow significantly. This is jus a base, 'usb_if_container' will be reworked anyway. I treat 'topic-usb' branch as wip/devel branch.

emob-nordic replied:
Yes. I don't want to make everything in one PR, as it will grow significantly. This is jus a base, 'usb_if_container' will be reworked anyway. I treat 'topic-usb' branch as wip/devel branch.

jfischer-phytec-iot commented on subsys/usb/usb_device.c:
```diff
@@ -1212,13 +1212,13 @@ static int class_handler(struct usb_setup_packet *pSetup,
 	size_t size = (__usb_data_end - __usb_data_start);
 	const struct usb_if_descriptor *if_descr;
-	struct usb_interface_cfg_data *iface;
+	const struct usb_interface_cfg_data *handlers;
```
const struct usb_request_handlers *handlers;

jfischer-phytec-iot commented on subsys/usb/usb_device.c:
```diff
@@ -1242,13 +1242,13 @@ static int custom_handler(struct usb_setup_packet *pSetup,
 	size_t size = (__usb_data_end - __usb_data_start);
 	const struct usb_if_descriptor *if_descr;
-	struct usb_interface_cfg_data *iface;
+	const struct usb_interface_cfg_data *handlers;
```
const struct usb_request_handlers *handlers;

jfischer-phytec-iot commented on subsys/usb/usb_device.c:
```diff
@@ -1275,7 +1276,7 @@ static int vendor_handler(struct usb_setup_packet *pSetup,
 {
 	size_t size = (__usb_data_end - __usb_data_start);
-	struct usb_interface_cfg_data *iface;
+	const struct usb_interface_cfg_data *handlers;
```
const struct usb_request_handlers *handlers;

jfischer-phytec-iot commented on include/usb/usb_device.h:
```diff
@@ -186,25 +186,28 @@ struct usb_request_handlers {
 	struct usb_request_handlers request_handlers;
-	/* Number of interfaces for this function */
+	/** Number of interfaces for this class */
 	const uint8_t num_interfaces;
```
const uint8_t num_interfaces; can be removed? or moved to usb_if_container?

emob-nordic replied:
`const uint8_t num_interfaces;` is number of `usb_if_containers` so it must stay here.

jfischer-phytec-iot replied:
Would it make sense to rename it too, `num_if_containers`?

emob-nordic replied:
Definitely yes. I will rename it. Wondering if We should also rethink the `usb_ep_cfg_data` and make it consistent with each other.

jfischer-phytec-iot replied:
Maybe just usb_ep_data or usb_ep_cfg?

emob-nordic replied:
I will leave endpoints for now. What I will change is:
`usb_cfg_data::interfaces -> usb_cfg_data::if_containers`
`usb_cfg_data::num_interfaces -> usb_cfg_data::num_if_containers`

jfischer-phytec-iot commented on include/usb/usb_device.h:
```diff
@@ -166,16 +180,13 @@ struct usb_request_handlers {
 
 /**
- * @brief USB device configuration
+ * @brief USB Class data
```
I guess this and L185:L187 should be part of the subsequent patch.

emob-nordic replied:
Yes. I will rebase it

Commit:
usb: Rework the USB core internals.

Rework the way how the interfaces are handled by the USB core stack.
Until now USB stack was aware of only the first interface associated
with each class. This led to a problem when not first interface was
addressed by the request.

This patch allow to filter out the request at core level.

The usb_cfg_data struct is legacy of initial implementation
of USB device stack. Initially the USB device stack was
implemented to support single class devices with
no composite support. This structure was intended to be used to
configure the device in run time by calling usb_set_config().

Because of how things grow the structure was used to represent
the classes instead of devices.

This patch removes device configuration descriptor pointer.
Adds pointer to table of USB interface containers associated with
he function it represents. Each container consist of pointer to
interface descriptor, its alternate setting (if exists) and currently
selected alternate setting. This help USB core stack to properly address
SET/GET interface requests.

This patch should be followed by renaming the structure to something
closer to 'class' specific data as described in USB 2.0 spec.

This patch is first step to rework the USB core stack and is not meant
to be considered as final form of USB core.

Signed-off-by: Emil Obalski <emil.obalski@nordicsemi.no>

carlescufi reviewed (approved):
This all looks very reasonable to me, including the new names.
